### PR TITLE
[etcd] Initial integration

### PR DESCRIPTION
**Pull request by AdamKorcz**
Initial integration of [etcd](https://github.com/etcd-io/etcd).
etcd is a consistent and highly-available key value store used as Kubernetes' backing store for all cluster data.

_____________________________________________________________

@ptabor Are you interested in integrating etcd into OSS-fuzz for continuous fuzzing? This PR adds the build files and the first fuzzer.
To complete the PR an email address is needed for bug reports.

### CODE REVIEW

File: projects/etcd/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2021 Google LLC
```
Comment (ptabor) [inline projects/etcd/Dockerfile:1]: The etcd Authors

File: projects/etcd/wal_fuzzer.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2021 Google LLC
```
Comment (ptabor) [inline projects/etcd/wal_fuzzer.go:1]: The etcd Authors

File: projects/etcd/wal_fuzzer.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	p, err := ioutil.TempDir("/tmp", "waltest")
+	if err != nil {
+		return -1
+	}
+	defer os.RemoveAll(p)
+	w, err := Create(zap.NewExample(), p, data)
+	if err != nil {
+		return 0
+	}
+	defer w.Close()
```
Comment (ptabor) [inline projects/etcd/wal_fuzzer.go:35]: Probably this should get paired with wal.ReadAll(...) that makes sure the same metadata are retrieved.

**Review by ptabor — COMMENTED** [review]
Thank you. 

I'm wondering whether for etcd fusing in the grpc layer (instead of unit-style aspects) would not be more valuable.   The code that is in backend (like wal) might have many 'implicit' assumptions (I'm not saying that's good, but that's the reality) that does not determine issues. 

But if there are data that can 'corrupt' the the public API surface - the problem starts to be significant.

File: projects/etcd/Dockerfile (hunk 1 above)
Comment (AdamKorcz) [reply projects/etcd/Dockerfile:1]: The files in this repo [must have the OSS-fuzz copyright header](https://google.github.io/oss-fuzz/getting-started/new-project-guide/#copyright-headers), however, once etcd has been integrated we can move both the fuzzer and build script upstream with amended copyright headers.

File: projects/etcd/wal_fuzzer.go (hunk 2 above)
Comment (AdamKorcz) [reply projects/etcd/wal_fuzzer.go:1]: See above

File: projects/etcd/wal_fuzzer.go (hunk 3 above)
Comment (AdamKorcz) [reply projects/etcd/wal_fuzzer.go]: Updated the fuzzer for roundtrip-testing.

File: projects/etcd/wal_fuzzer.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	if err != nil {
+		return 0
+	}
+	defer neww.Close()
+	metadata, _, _, err := neww.ReadAll()
+	if err != nil {
+		return 0
+	}
+	if !bytes.Equal(data, metadata) {
+		panic("data and metadata are not similar, but they should be")
```
Comment (ptabor) [inline projects/etcd/wal_fuzzer.go:55]: similar -> equal

File: projects/etcd/Dockerfile (hunk 1 above)
Comment (ptabor) [reply projects/etcd/Dockerfile:1]: I'm sorry. Started to read this with etcd-repo mindset.

**Comment by ptabor** [comment]
Thank you: You can send the initial reports to me: ptab@google.com and later we will think how to expand this to wider community. 

Can the e-mail be a google-group ?

**Comment by AdamKorcz** [comment]
> Can the e-mail be a google-group ?
For notifications yes, but access to bug reports is not granted to members of the group.

**Comment by AdamKorcz** [comment]
@DavidKorczynski @jonathanmetzman @oliverchang @inferno-chromium @asraa etcd is ready for review.

**Review by oliverchang — APPROVED** [review]

**Pull request merged**